spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +70,97 @@ def __iter__(self):
                 self._add_chunk_to_size(chunk)
                 yield chunk
 
-    def _add_chunk_to_size(self, chunk):
+    def _add_chunk_to_size(self, chunk) -> None:
         self.total_size += len(chunk)
-        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+        if self.total_size > self._maximum_allowed_size:
             raise ArchiveTooLargeError()
 
-    def read(self, size=-1) -> bytes:
-        content = self._trailing_data
-        self._trailing_data = b''
-
-        if not size or size == -1:
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-        else:
-            # iterate chunks until we've read at least `size` amount of data
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-                if len(content) > size:
-                    break
+    def _num_bytes_left_in_buffer(self) -> int:
+        return len(self._buffer) - self._buffer_position
+
+    def _should_read_more(self, size) -> bool:
+        return (not size or self._num_bytes_left_in_buffer() < size) and not self._consumed
+
+    def read(self, size=-1) -> memoryview:
```

Comment:
Simplified the implementation here a bit based on feedback - we also now only ever return a `memoryview` from this